lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp

```diff
@@ -406,9 +407,227 @@ struct ConvolutionToConv2dPattern
+        generateTransposeIndicesRightToLeft(desired_layout, current_layout);
+
+    auto kernel_height_attr = rewriter.getSI32IntegerAttr(
+        static_cast<int32_t>(op.getWindowDimensions()[spatial_dims[0]]));
+    auto kernel_width_attr = rewriter.getSI32IntegerAttr(
+        static_cast<int32_t>(op.getWindowDimensions()[spatial_dims[1]]));
+
+    auto stride_height_attr = rewriter.getSI32IntegerAttr(
+        static_cast<int32_t>(op.getWindowStrides()[spatial_dims[0]]));
+
+    auto stride_width_attr = rewriter.getSI32IntegerAttr(
+        static_cast<int32_t>(op.getWindowStrides()[spatial_dims[1]]));
+
+    auto dilation_height_attr = rewriter.getSI32IntegerAttr(
+        adaptor.getWindowDilations()[spatial_dims[0]]);
+    auto dilation_width_attr = rewriter.getSI32IntegerAttr(
+        adaptor.getWindowDilations()[spatial_dims[1]]);
+    auto ceil_mode_attr = rewriter.getBoolAttr(false);
+
+    auto padding_top_attr =
+        rewriter.getSI32IntegerAttr(op.getPadding()[2 * spatial_dims[0]]);
+    auto padding_bottom_attr =
+        rewriter.getSI32IntegerAttr(op.getPadding()[2 * spatial_dims[0] + 1]);
+    auto padding_left_attr =
+        rewriter.getSI32IntegerAttr(op.getPadding()[2 * spatial_dims[1]]);
+    auto padding_right_attr =
+        rewriter.getSI32IntegerAttr(op.getPadding()[2 * spatial_dims[1] + 1]);
+    auto operand_constraints = adaptor.getOperandConstraints();
+
+    std::vector<Value> outputs;
+    for (Value input : adaptor.getInputs()) {
+      input = generateTransposeSequence(
+          input, rewriter, input_transpose_indices, operand_constraints);
+
+      auto outputType = mlir::cast<RankedTensorType>(op.getResult(0).getType());
+      auto newOutputShape = outputType.getShape().vec();
+      for (TransposeDims dims : input_transpose_indices) {
+        std::swap(newOutputShape[std::get<0>(dims)],
+                  newOutputShape[std::get<1>(dims)]);
+      }
+      auto newOutputType =
+          outputType.cloneWith(newOutputShape, outputType.getElementType());
+      auto outputTensor = rewriter.create<tensor::EmptyOp>(
+          op.getLoc(), newOutputType.getShape(),
+          newOutputType.getElementType());
+
+      auto new_pool = rewriter.create<PoolOpType>(
+          op.getLoc(), newOutputType, input, outputTensor, kernel_height_attr,
+          kernel_width_attr, stride_height_attr, stride_width_attr,
+          dilation_height_attr, dilation_width_attr, ceil_mode_attr,
+          padding_top_attr, padding_bottom_attr, padding_left_attr,
+          padding_right_attr, operand_constraints);
+
+      Value output = generateTransposeSequence(new_pool.getResult(), rewriter,
+                                               output_transpose_indices,
+                                               operand_constraints);
+      outputs.push_back(output);
+    }
+
+    for (int64_t i = 0; i < static_cast<int64_t>(outputs.size()); i++) {
+      rewriter.replaceAllUsesWith(op.getResult(i), outputs[i]);
```

Comment:
`replaceAllUsesWith` should not be used in dialect conversion. It's a bug that it's available to dialect conversion's rewriter. It modifies IR, which should not be allowed in dialect conversion rewriter.

Reply:
Not sure what to do here then. The `PoolingOp` can perform multiple pools at once. That is, you could pass it some number (lets say 5) tensors, pool all of them, and then get 5 results. As part of this conversion I am splitting that up into individual MaxPool2d ops. So I need to replace the use of the individual results with each new MaxPool2d op.

Also this isn't dialect conversion its just a decomposition pass within the same dialect.

Reply:
Looks like this PR will solve this issue for me actually so don't worry about my floating-point bit-shifting hack: https://github.com/tenstorrent/tt-mlir/pull/1164#pullrequestreview-2424129767

Reply:
I've resolved this. I didn't realize I could use `replaceOp` with an op and range of results.

Reply:
> Also this isn't dialect conversion its just a decomposition pass within the same dialect.

True, but we are actually using dialect conversion for the decomp, as it offers op legality checks on target IR.